vllm/engine/metrics.py

```diff
@@ -81,6 +97,18 @@ def __init__(self, labelnames: List[str]):
             documentation="Histogram of end to end request latency in seconds.",
             labelnames=labelnames,
             buckets=[1.0, 2.5, 5.0, 10.0, 15.0, 20.0, 30.0, 40.0, 50.0, 60.0])
+        self.histogram_max_tokens = Histogram(
+            name="vllm:request_params_max_tokens",
+            documentation="Histogram of the max_tokens request parameter.",
+            labelnames=labelnames,
+            buckets=build_1_2_5_buckets(max_model_len),
+        )
+        self.histogram_request_n = Histogram(
+            name="vllm:request_params_n",
+            documentation="Histogram of the n request parameter.",
+            labelnames=labelnames,
+            buckets=[1, 2, 5, 10, 20],
+        )
```

Comment:
Maybe these should go in an `Info`, like the `cache_config`?

Reply:
IIUC, `Info` is intended to collect constant metrics such as configuration values. `vllm:request_params_max_tokens` and `vllm:request_params_n` are intended to collect the values of the `max_tokens` and `n` arguments in each request, which may vary with each request.
For instance, in the following request, `max_tokens=7` and `n=3` will be collected.
```
curl http://localhost:8000/v1/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "facebook/opt-125m",
        "prompt": "San Francisco is a",
        "max_tokens": 7,
        "best_of": 5,
        "n": 3,
        "use_beam_search": "true",
        "temperature": 0
    }'
```

Reply:
Good point.

Although, I'm not convinced that it's a useful datapoint to collect (I am open to being convinced though!). i.e. what do we, as the vLLM provider, learn from the token limit the user sets?

Also, in the vast majority of use cases a user will statically set their sampling parameters in their application and then never touch them again.

Reply:
I see your point. Having a histogram of the token limit set by the user might indeed be redundant, particularly considering we have a histogram of the actual number of generated tokens per request (`vllm:request_generation_tokens_sum`). I'll remove it.


Regarding `n` and `best_of`, I realized I overlooked adding a metric for collecting `best_of`. Just to clarify, `best_of` determines the width of the beam search, while `n` specifies how many "top beams" to return (`n` <= `best_of`). AFAIU, larger values of `best_of` can significantly impact the engine, as the batch will be dominated by sequences from a few requests.

It might be insightful to compare the histograms of both `n` and `best_of`. A significant deviation between them could suggest that the engine is processing a substantial number of tokens that users aren't actually consuming.

Please let me know your thoughts on this.

Reply:
Knowing the amount of tokens "wasted" in beam search could be interesting to know as a developer of vLLM. i.e. if beam search is being used extensively and many tokens are being "wasted", it signals that we need to optimise beam search if we can.

@simon-mo @Yard1 what do you think about this?